File: src/embodied-intelligence/src/3D_human_pose_recognition/ios_sdk/KeyPointCenter/KeyPoint/ConstantVelocity3DModel.swift
//
//  ConstantVelocity3DModel.swift
//  MetaScripts
//
//  Created by ZhaoTianyu on 2024/7/8.
//

import Foundation
import simd

class ConstantVelocity3DModel: NSObject, NSCopying {
    typealias StateConvertFunc = (ConstantVelocity3DModel) -> [Float]
    typealias StateConvertBackFunc = (VectorA) -> ConstantVelocity3DModel
    
    // Fields
    static let Dimension = 6
    var Position = simd_float3()
    var Velocity = simd_float3()

    // Methods
    func copy(with zone: NSZone? = nil) -> Any {
        let copy = ConstantVelocity3DModel()
        copy.Position = self.Position
        copy.Velocity = self.Velocity
        return copy
    }
    
    static var fromArray: StateConvertBackFunc = {
        let model = ConstantVelocity3DModel()
        model.Position = simd_float3($0[0], $0[2], $0[4])
        model.Velocity = simd_float3($0[1], $0[3], $0[5])
        return model
    }

    static func getPositionMeasurementMatrix() -> [[Float]] {
        return [
            [1, 0, 0, 0, 0, 0],
            [0, 0, 1, 0, 0, 0],
            [0, 0, 0, 0, 1, 0]
        ]
    }

    static func getProcessNoise(_ accelerationNoise: Float, timeInterval: Float = 1) -> [[Float]] {
        var matrix = [[Float]](repeating: [Float](repeating: 0, count: 3), count: 6)
        matrix[0][0] = (timeInterval * timeInterval) / 2
        matrix[1][0] = timeInterval
        matrix[2][1] = (timeInterval * timeInterval) / 2
        matrix[3][1] = timeInterval
        matrix[4][2] = (timeInterval * timeInterval) / 2
        matrix[5][2] = timeInterval

        let b = MatrixF.diagonal(columnCount: matrix[0].count, value: accelerationNoise)
        return MatrixF.mul(MatrixF.mul(matrix, b), MatrixF.transpose(matrix))
    }

    static func getTransitionMatrix(timeInterval: Float = 1) -> [[Float]] {
        let num = timeInterval
        var matrix = [[Float]](repeating: [Float](repeating: 0, count: 6), count: 6)
        matrix[0][0] = 1
        matrix[1][1] = 1
        matrix[2][2] = 1
        matrix[3][3] = 1
        matrix[4][4] = 1
        matrix[5][5] = 1
        matrix[0][1] = num
        matrix[2][3] = num
        matrix[4][5] = num
        return matrix
    }
    
    static var toArray: StateConvertFunc = { modelState in
        return [modelState.Position.x, modelState.Velocity.x, modelState.Position.y, modelState.Velocity.y, modelState.Position.z, modelState.Velocity.z]
    }
}

File: src/embodied-intelligence/src/3D_human_pose_recognition/ios_sdk/UnityDEBUG/ForDebug/Assets/Scripts/CharacterDriven/ConstantVelocity3DModel.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class ConstantVelocity3DModel : ICloneable
{
    // Fields
    public const int Dimension = 6;
    public Vector3 Position = Vector3.zero;
    public Vector3 Velocity = Vector3.zero;

    // Methods
    public object Clone()
    {
        return new ConstantVelocity3DModel { Position = this.Position, Velocity = this.Velocity };
    }

    public static ConstantVelocity3DModel FromArray(float[] arr)
    {
        return new ConstantVelocity3DModel { Position = new Vector3(arr[0], arr[2], arr[4]), Velocity = new Vector3(arr[1], arr[3], arr[5]) };
    }

    public static float[,] GetPositionMeasurementMatrix()
    {
        return new float[,] { { 1f, 0f, 0f, 0f, 0f, 0f }, { 0f, 0f, 1f, 0f, 0f, 0f }, { 0f, 0f, 0f, 0f, 1f, 0f } };
    }

    public static float[,] GetProcessNoise(float accelerationNoise, float timeInterval = 1f)
    {
        float[,] singleArray1 = new float[6, 3];
        singleArray1[0, 0] = (timeInterval * timeInterval) / 2f;
        singleArray1[1, 0] = timeInterval;
        singleArray1[2, 1] = (timeInterval * timeInterval) / 2f;
        singleArray1[3, 1] = timeInterval;
        singleArray1[4, 2] = (timeInterval * timeInterval) / 2f;
        singleArray1[5, 2] = timeInterval;
        float[,] matrix = singleArray1;
        float[,] b = MatrixF.Diagonal<float>(matrix.ColumnCount<float>(), accelerationNoise);
        return matrix.Multiply(b).Multiply(matrix.Transpose<float>());
    }

    public static float[,] GetTransitionMatrix(float timeInterval = 1f)
    {
        float num = timeInterval;
        float[,] singleArray1 = new float[,] { { 1f, 0f, 0f, 0f, 0f, 0f }, { 0f, 1f, 0f, 0f, 0f, 0f }, { 0f, 0f, 1f, 0f, 0f, 0f }, { 0f, 0f, 0f, 1f, 0f, 0f }, { 0f, 0f, 0f, 0f, 1f, 0f }, { 0f, 0f, 0f, 0f, 0f, 1f } };
        singleArray1[0, 1] = num;
        singleArray1[2, 3] = num;
        singleArray1[4, 5] = num;
        return singleArray1;
    }

    public static float[] ToArray(ConstantVelocity3DModel modelState)
    {
        return new float[] { modelState.Position.x, modelState.Velocity.x, modelState.Position.y, modelState.Velocity.y, modelState.Position.z, modelState.Velocity.z };
    }
}


